src/Ocelot.Provider.Polly/PollyQoSProvider.cs

```diff
@@ -1,54 +1,81 @@
+        HttpStatusCode.VariantAlsoNegotiates,
+        HttpStatusCode.InsufficientStorage,
+        HttpStatusCode.LoopDetected,
+    };
+
+    public PollyQoSProvider(IOcelotLoggerFactory loggerFactory)
+    {
+        _logger = loggerFactory.CreateLogger<PollyQoSProvider>();
+    }
+
+    private static string GetRouteName(DownstreamRoute route)
+    {
+        return string.IsNullOrWhiteSpace(route.ServiceName)
+            ? route.UpstreamPathTemplate?.Template ?? route.DownstreamPathTemplate?.Value ?? string.Empty
+            : route.ServiceName;
+    }
+
+    public CircuitBreaker<HttpResponseMessage> GetCircuitBreaker(DownstreamRoute route)
+    {
+        lock (_lockObject)
+        {
+            var currentRouteName = GetRouteName(route);
+            if (!_circuitBreakers.ContainsKey(currentRouteName))
             {
-                var info = $"Route: {GetRouteName(route)}; Breaker logging in {nameof(PollyQoSProvider)}: ";
-                var logger = loggerFactory.CreateLogger<PollyQoSProvider>();
-                circuitBreakerPolicy = Policy
-                    .Handle<HttpRequestException>()
-                    .Or<TimeoutRejectedException>()
-                    .Or<TimeoutException>()
-                    .CircuitBreakerAsync(
-                        exceptionsAllowedBeforeBreaking: route.QosOptions.ExceptionsAllowedBeforeBreaking,
-                        durationOfBreak: TimeSpan.FromMilliseconds(route.QosOptions.DurationOfBreak),
-                        onBreak: (ex, breakDelay) =>
-                            logger.LogError(info + $"Breaking the circuit for {breakDelay.TotalMilliseconds} ms!", ex),
-                        onReset: () =>
-                            logger.LogDebug(info + "Call OK! Closed the circuit again."),
-                        onHalfOpen: () =>
-                            logger.LogDebug(info + "Half-open; Next call is a trial.")
-                    );
-            }
-
-            _ = Enum.TryParse(route.QosOptions.TimeoutStrategy, out TimeoutStrategy strategy);
-            var timeoutPolicy = Policy.TimeoutAsync(TimeSpan.FromMilliseconds(route.QosOptions.TimeoutValue), strategy);
-            CircuitBreaker = new CircuitBreaker(circuitBreakerPolicy, timeoutPolicy);
+                _circuitBreakers.Add(currentRouteName, CircuitBreakerFactory(route));
+            }
+
+            return _circuitBreakers[currentRouteName];
         }
+    }
+
+    private CircuitBreaker<HttpResponseMessage> CircuitBreakerFactory(DownstreamRoute route)
+    {
+        AsyncCircuitBreakerPolicy<HttpResponseMessage> exceptionsAllowedBeforeBreakingPolicy = null;
+        if (route.QosOptions.ExceptionsAllowedBeforeBreaking > 0)
+        {
+            var info = $"Route: {GetRouteName(route)}; Breaker logging in {nameof(PollyQoSProvider)}: ";
+
+            exceptionsAllowedBeforeBreakingPolicy = Policy
+                .HandleResult<HttpResponseMessage>(r => _serverErrorCodes.Contains(r.StatusCode))
```

Comment:
Cannot understand that: `_serverErrorCodes` ...
Why do we handle all these HTTP statuses?
Is that some kind of best practice for Polly?

Reply:
That was an assumption. Http Status is returned from SendAsync, polly will react on certain types. I thought 500...508 is great.

Reply:
If internal Polly implementations use these status codes (or overrides them) then it is OK for me to process them.

Reply:
Yes it reacts on the status codes you choose, or other specs from HttpResponseMessage class... With that we can avoid the overhead of throwing an exception.